rabbitmq/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/sinks/RabbitMQSink.scala

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (c) 2013-2022 Snowplow Analytics Ltd. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package com.snowplowanalytics.snowplow.collectors.scalastream
+package sinks
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.concurrent.duration._
+import scala.util.{Failure, Success}
+
+import cats.syntax.either._
+
+import com.rabbitmq.client.Channel
+
+import com.snowplowanalytics.snowplow.collectors.scalastream.model.RabbitMQBackoffPolicyConfig
+
+class RabbitMQSink(
+  channel: Channel,
+  exchangeName: String,
+  backoffPolicy: RabbitMQBackoffPolicyConfig
+) extends Sink {
+
+  implicit val ec = ExecutionContext.global
```

Comment:
Shouldn't we be running sink in its own thread pool?
Given there is a slowdown on AMQP messaging it might clog the main thread pool making the collector unresponsive. 
I believe we should stick it into its own (probably fixed size) thread pool to avoid it interfering with HTTP layer.

Reply:
Very good point! I had actually thought of doing that, and didn't do it because I didn't know how to decide the size of this thread pool (I'm ashamed). What would you recommend?

Reply:
Nothing to be ashamed of! This is a really tricky topic and the consequences are unobvious. It is also very much dependent on runtime environment so I believe it should be configurable by the user. And optional. As a default, in terms of how we determine the thread pool size I think [this](https://engineering.zalando.com/posts/2019/04/how-to-set-an-ideal-thread-pool-size.html) summarises in pretty well. I've followed the Goetz approach as a starting point for when I had to set these things up before. And it usually works for usual cases. But I think it's a perf tuning thing, and just sticking it into another threadpool of some size will make things much more resilient. Perf improvements probably come in later especially as we're introducing entirely new transport layer.

Reply:
Nothing to be ashamed of! This is a tricky one. Especially as it's very much runtime-dependent. I'd expose an optional setting for this. Now, about how we set the default it probably the interesting part. I think [this](https://engineering.zalando.com/posts/2019/04/how-to-set-an-ideal-thread-pool-size.html) is a pretty useful summary.
All in all I've usually followed the Goetz's approach and as a starting point it served well. Now, as we're introducing a new transport layer we will do lots of performance testing and tuning along the way. I'd start with some rough calculation using the `Runtime.getRuntime().availableProcessors()` and we'll be able to override that if necessary.

Reply:
About `Runtime.getRuntime().availableProcessors()`:
```
But be careful with this command if you run your service in a container environment. Without specifying any constraints, a containerized process will be able to see the hardware on the host OS.
```

I find it super hard to decide a default size, especially because it depends mostly about RabbitMQ and collector can't know that. So here is what I propose: make thread pool size configurable and optional. If it is set, use it for the fixed pool size. If it's not there, use a cached thread pool.

In both cases, we can't avoid an OOM if RabbitMQ goes down because the collector will keep accepting new requests. In both cases, RabbitMQ won't get flooded with requests when it's up again because all the requests go through the same channel.

WDYT?

Reply:
> About Runtime.getRuntime().availableProcessors():

I think it's been like this prior to JDK11. It introduced `-XX-UseContainerSupport` (as a non-default) which fixed a bunch of resource issues for containers. JDK17 fixes even more. But yes, totally depends on the runtime *we pick*. We had to set flags on containers and even then still, java lied about what's available to the container because it wasn't aware of the cgroups (v2!) settings.

So, the behaviour is: if you run java on a host that only supports groups v2 and not groups v1, you will get the host's resources. Meaning, we need to consider that.

> make thread pool size configurable and optional. If it is set, use it for the fixed pool size. If it's not there, use a cached thread pool.
Makes sense. We totally need to document the runtime and consider the JDK platform we use for build/runtime.